### PR TITLE
Enhance Docket Model to better mirror PACER

### PR DESCRIPTION
**Pull request by flooie**
We need to make some changes to the docket model to include more granular docket number data. 

This includes adding a few fields that are captured partially in docket number 

- office_code
- case type 

and brand new fields 

- defendant_number 
- judge_initials

and finally adding 

- parent_docket

a self-referential foreign key relationship within the docket model that will enable parent/child relationship of dockets with more than one defendant.  This is needed so we can begin to group dockets together in the same manner that pacer does it

### CODE REVIEW

**Comment by johnhawkinson** [comment]
>  same manner that pacer does it

PACER? What's PACER? I'm sure ACMS doesn't do it this way. I think you mean CMECF ;)
Anyhow:

```
         migrations.AddField(
            model_name="docket",
            name="judge_initials",
            field=models.CharField(
                blank=True,
                help_text="A typically three-letter upper cased abbreviation of the judge's initials. In the example 2:07-cv-34911-MJL, MJL is the judge's initials. Judge initials change if a new judge takes over a case.",
                max_length=4,
            ),
        ),
```

Some districts (EDNY) use 3 initials for article III judges and another 3 after that for magistrate judges. Is the intention that the latter should be ignored? It is not 100% clear from the `help_text`.

Also, some districts (I thought SDFL but no…somewhere near htere?) use a secret decoder ring of codes like `-A` or `-AA` to represent juges, but they are not initials. Again, the `help_text` guidance is not clear.

Best to fix these things sooner.

File: cl/search/models.py
Comment (mlissner) [inline cl/search/models.py:587]: I think you want blank=True on this too, no?

File: cl/search/models.py
Comment (mlissner) [inline cl/search/models.py:573]: Do we want an index on this? If so, can you put it in the meta class?

File: cl/search/models.py
Comment (mlissner) [inline cl/search/models.py:1]: Looking good. A few thoughts:

 - We have [a spreadsheet of office codes](https://free.law/xlsx/fjc/integrated-database/office-codes.xlsx). They go from 0-9 and from A-R, I think, but none is more than a single character. 

    Since this table is used for state cases as well, should we call this `federal_dn_office_code` to make it clear that it doesn't apply to state cases? And we can make sure to use the words "docket number" in the description so people know what dn stands for?

 - For `case_type` feels like we should lower- or uppercase it just as a normalization step. I think I usually see it lowercase, so maybe it's worth doing that (though Django doesn't make this easy). 

    Again, maybe something like `federal_dn_case_type` would be better? I'm not sure I like `case_type` because it feels like it's over promising, but maybe it's fine.

 - For judge initials, I know those initials!! But also, I agree with John that we should allow more length in here. It makes sense to me that if courts are doing inconsistent things, we should just dump it all in here. That means it will need more length.

    And again, perhaps `federal_dn_judge_initials`? 

 - Finally, `defendant_number` could be `federal_defendant_number`?

I'm not sure if `dn` is better than doing `docket_number`. Probably not, actually, but I thought it would be worth trying on.

**Review by mlissner — COMMENTED** [review]
A few comments. The one other thing I'm thinking about is that it'd be nice to do this one and the ordering one together, because I think it's going to be a race to see which one gets in first, and then deployment will be twice as hard if they're separate.

Thanks!

File: cl/search/models.py
Comment (flooie) [reply cl/search/models.py:573]: good question - @ERosendo and I had a long back and forth before we removed all the indexes I had originally proposed.

File: cl/search/models.py
Comment (flooie) [reply cl/search/models.py:587]: Yes I think that is right.

**Comment by flooie** [comment]
@johnhawkinson you make a good point - I think we would want to capture the magistrate initials as well.  Might as well complete the collection @mlissner what do you think?

File: cl/search/models.py
Comment (mlissner) [inline cl/search/models.py:564]: John makes the case that office codes should be more than one character and sure, why not. Want to make them three to provide some headroom?

File: cl/search/models.py
Comment (flooie) [reply cl/search/models.py:1]: I was hesitant to normalize the field because the system requires the correct capitalization.  Its true that we only see the upper case case types in one district PAED but you can see the capitalization is required for the the possible_case_numbers api query

`<request number="16al27244"/>`

```
<request number="16AL27244">
<case number="2:16-AL-27244" id="534078" title="2:16-AL-27244-CMR ALBUTEROL CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-AL-27244-CMR"/>
</request>
```

Comment (flooie) [reply cl/search/models.py:1]: I didn't think @johnhawkinson was advocating for longer fields in judge initials, I think we should just add the magistrate initials as a field as well.  

```
<request number="1627244">
<case number="2:16-AL-27244" id="534078" title="2:16-AL-27244-CMR ALBUTEROL CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-AL-27244-CMR"/>
<case number="2:16-AM-27244" id="534079" title="2:16-AM-27244-CMR AMITRIPTYLINE CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-AM-27244-CMR"/>
<case number="2:16-BC-27244" id="534080" title="2:16-BC-27244-CMR BACLOFEN CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-BC-27244-CMR"/>
<case number="2:16-BZ-27244" id="534081" title="2:16-BZ-27244-CMR BENAZEPRIL HCTZ CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-BZ-27244-CMR"/>
<case number="2:16-CB-27244" id="534082" title="2:16-CB-27244-CMR CLOBETASOL CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-CB-27244-CMR"/>
<case number="2:16-CM-27244" id="534083" title="2:16-CM-27244-CMR CLOMIPRAMINE CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-CM-27244-CMR"/>
<case number="2:16-DG-27244" id="534085" title="2:16-DG-27244-CMR DIGOXIN CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-DG-27244-CMR"/>
<case number="2:16-DS-27244" id="534084" title="2:16-DS-27244-CMR DESONIDE CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-DS-27244-CMR"/>
<case number="2:16-DV-27244" id="534086" title="2:16-DV-27244-CMR DIVALPROEX ER CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-DV-27244-CMR"/>
<case number="2:16-DX-27244" id="534087" title="2:16-DX-27244-CMR DOXYCYCLINE CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-DX-27244-CMR"/>
<case number="2:16-EC-27244" id="534088" title="2:16-EC-27244-CMR ECONAZOLE NITRATE CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-EC-27244-CMR"/>
<case number="2:16-FL-27244" id="534089" title="2:16-FL-27244-CMR FLUOCINONIDE CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-FL-27244-CMR"/>
<case number="2:16-GL-27244" id="534090" title="2:16-GL-27244-CMR GLYBURIDE CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-GL-27244-CMR"/>
<case number="2:16-LD-27244" id="534092" title="2:16-LD-27244-CMR LIDOCAINE/PRILOCAINE CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-LD-27244-CMR"/>
<case number="2:16-LV-27244" id="534091" title="2:16-LV-27244-CMR LEVOTHYROXINE CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-LV-27244-CMR"/>
<case number="2:16-PP-27244" id="534094" title="2:16-PP-27244-CMR PROPRANOLOL CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-PP-27244-CMR"/>
<case number="2:16-PV-27244" id="534093" title="2:16-PV-27244-CMR PRAVASTATIN CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-PV-27244-CMR"/>
<case number="2:16-UR-27244" id="534095" title="2:16-UR-27244-CMR URSODIOL CASES (STATE ATTORNEYS GENERAL) (closed 06/05/2018)" sortable="2:2016-UR-27244-CMR"/>
</request>
```

also I found a jackpot of capitalized case types

Comment (mlissner) [reply cl/search/models.py:1]: I don't know what John was suggesting, but I guess I'd like to do whatever PACER does internally. If they have a field for mag initials, fine. If they don't, and one court is just doing weird things, then I think we shouldn't try to have our own field.

Comment (mlissner) [reply cl/search/models.py:1]: Agreed about not normalizing case letters. Who knows what's going on inside PACER with these things. Maybe it matters, maybe not, but until we know for sure that lowercase is safe, let's not make assumptions.

File: cl/search/models.py
Comment (flooie) [reply cl/search/models.py:573]: @mlissner, @ERosendo and I were discussing whether the low cardinality of the field would actually hurt in this situation.  Considering CR and CV MJ and BK are so much larger than other case types we left it off.  But we would be open to hearing your thoughts.

**Comment by flooie** [comment]
> > same manner that pacer does it
> 
> PACER? What's PACER? I'm sure ACMS doesn't do it this way. I think you mean CMECF ;) Anyhow:
> 
> ```
>          migrations.AddField(
>             model_name="docket",
>             name="judge_initials",
>             field=models.CharField(
>                 blank=True,
>                 help_text="A typically three-letter upper cased abbreviation of the judge's initials. In the example 2:07-cv-34911-MJL, MJL is the judge's initials. Judge initials change if a new judge takes over a case.",
>                 max_length=4,
>             ),
>         ),
> ```
> 
> Some districts (EDNY) use 3 initials for article III judges and another 3 after that for magistrate judges. Is the intention that the latter should be ignored? It is not 100% clear from the `help_text`.
> 
> Also, some districts (I thought SDFL but no…somewhere near htere?) use a secret decoder ring of codes like `-A` or `-AA` to represent juges, but they are not initials. Again, the `help_text` guidance is not clear.
> 
> Best to fix these things sooner.

<case number="1:01-cv-570" id="75595" title="1:01-cv-00570-PCH Wohlfarth v. City of Miami, et al (closed 11/19/2001)" sortable="1:2001-cv-00570-PCH"/>

This looks like it stands [Senior Judge Paul C. Huck] - perhaps though it's a different district or something they do elsewhere.  I havent seen any examples where they dont correspond to the judge initials

File: cl/search/models.py
Comment (mlissner) [reply cl/search/models.py:573]: Cool.

File: cl/search/models.py
Comment (johnhawkinson) [reply cl/search/models.py:1]: What's this PACER thing? CMECF separates judge initials by a hyphen between primary judge initials and secondary/referral/magistrate judge initials, I would call those two fields, but you could call them one with an internal separator and two sub-fields. I would suggest two fields.

But what John was really suggesting was that the field description needs to be **clear and unambiguous**, and that it was not. He did not previously express a preference for how it should be structured.

**Comment by johnhawkinson** [comment]
I found an example of non-initial judge codes, https://www.ord.uscourts.gov/index.php/172-filing/help-and-training/filing-documents/337-case-codes-explained, which seem to be sometimes initials and sometimes first two letters of the surname:

```
SB 	Judge Stacie F. Beckerman
BR 	Judge Anna J. Brown
```

Not that it matters, but I still think there's at least one court that does them A/B/C/D/E/F but I can't find it easily.

**Comment by flooie** [comment]
@mlissner I think this this is ready for your review

**Review by mlissner — APPROVED** [review]

File: cl/api/templates/bulk-data.html
Comment (mlissner) [inline cl/api/templates/bulk-data.html:167]:
```suggestion
      <strong>2024-08-02</strong>: Add new fields to the bulk data files for the Docket object: federal_dn_case_type, federal_dn_office_code, federal_dn_judge_initials_assigned, federal_dn_judge_initials_referred, federal_defendant_number, parent_docket_id.
```

**Comment by mlissner** [comment]
Still need step 1.iv. here, please: https://github.com/freelawproject/courtlistener/wiki/Database-migrations#steps-to-making-new-migrations

**Review by mlissner — CHANGES_REQUESTED** [review]

**Review by mlissner — APPROVED** [review]

**Comment by mlissner** [comment]
LGTM, thanks @flooie. Merged! Deployment will fail until Ramiro takes manual action, but that can wait until #2821 is landed.

**Pull request merged**
